### PR TITLE
fix: let information_schema know itself

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

makes `SHOW DATABASES`/`information_schema.tables`/`information_schema.columns` know themselves.

There are some dumb if-else branches for `FrontendCatalogManager`.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by waynexia** [comment]
@zhongzc PTAL

File: tests/cases/standalone/common/system/information_schema.result [hunk 1]
```diff
@@ -31,25 +31,43 @@ where table_catalog = 'greptime'
...
-+---------------+--------------+------------+------------+--------+
-| greptime      | my_db        | foo        | BASE TABLE | mito   |
-+---------------+--------------+------------+------------+--------+
++---------------+--------------------+------------+-----------------+--------+
+| table_catalog | table_schema       | table_name | table_type      | engine |
++---------------+--------------------+------------+-----------------+--------+
+| greptime      | information_schema | columns    | LOCAL TEMPORARY |        |
+| greptime      | information_schema | tables     | LOCAL TEMPORARY |        |
+| greptime      | my_db              | foo        | BASE TABLE      | mito   |
++---------------+--------------------+------------+-----------------+--------+
```
Comment (zhongzc) [inline tests/cases/standalone/common/system/information_schema.result:40]: Considering the purpose of the test, just add a filter condition `table_schema != 'information_schema'`
```suggestion
  and table_schema != 'public'
  and table_schema != 'information_schema'
order by table_schema, table_name;

| table_catalog | table_schema | table_name | table_type | engine |
+---------------+--------------+------------+------------+--------+
| greptime      | my_db        | foo        | BASE TABLE | mito   |
+---------------+--------------+------------+------------+--------+
```

File: tests/cases/standalone/common/system/information_schema.result [hunk 2]
```diff
@@ -31,25 +31,43 @@ where table_catalog = 'greptime'
   and table_schema != 'public'
```
Comment (zhongzc) [inline tests/cases/standalone/common/system/information_schema.result:31]: Considering the purpose of the test, just add a filter condition `table_schema != 'information_schema'`
```suggestion
  and table_schema != 'public'
  and table_schema != 'information_schema'
```

File: tests/cases/standalone/common/system/information_schema.result [hunk 3]
```diff
@@ -31,25 +31,43 @@ where table_catalog = 'greptime'
...
++---------------+--------------------+------------+-----------------+--------+
+| greptime      | information_schema | columns    | LOCAL TEMPORARY |        |
+| greptime      | information_schema | tables     | LOCAL TEMPORARY |        |
+| greptime      | my_db              | foo        | BASE TABLE      | mito   |
++---------------+--------------------+------------+-----------------+--------+
 
 select table_catalog, table_schema, table_name, column_name, data_type, semantic_type
 from information_schema.columns
 where table_catalog = 'greptime'
   and table_schema != 'public'
```
Comment (zhongzc) [inline tests/cases/standalone/common/system/information_schema.result:45]: ditto

File: src/catalog/src/local/memory.rs [hunk 4]
```diff
@@ -260,6 +245,29 @@ impl CatalogManager for MemoryCatalogManager {
 }
 
 impl MemoryCatalogManager {
+    /// Yes this is the `default()` constructor
+    #[allow(clippy::should_implement_trait)]
+    pub fn default() -> Arc<Self> {
```
Comment (zhongzc) [inline src/catalog/src/local/memory.rs:250]: `default()` should not do too many things, may you consider moving to the another constructor

File: src/catalog/src/information_schema.rs [hunk 5]
```diff
@@ -51,42 +56,93 @@ impl InformationSchemaProvider {
...
     }
-}
 
-impl InformationSchemaProvider {
-    pub fn table(&self, name: &str) -> Result<Option<TableRef>> {
-        let stream_builder = match name.to_ascii_lowercase().as_ref() {
-            TABLES => Arc::new(InformationSchemaTables::new(
-                self.catalog_name.clone(),
-                self.catalog_manager.clone(),
+    pub fn build_information_schema(
```
Comment (zhongzc) [inline src/catalog/src/information_schema.rs:60]: Not sure but
```suggestion
    pub fn tables(
```

Comment (sunng87) [reply src/catalog/src/information_schema.rs:60]: `tables` makes sense in terms of its return value, comparing to the `table` which returns a single table. However input of these two functions varies, which indicates they operate at different levels.

I would suggest to use `build` as name.

**Review by sunng87 — APPROVED** [review]

File: src/catalog/src/information_schema/tables.rs [hunk 6]
```diff
@@ -151,21 +155,43 @@ impl InformationSchemaTablesBuilder {
...
+                            self.add_table(
+                                &catalog_name,
+                                &schema_name,
+                                &table_name,
+                                TableType::Temporary,
+                                Some(INFORMATION_SCHEMA_TABLES_TABLE_ID),
+                                None,
+                            );
+                        }
+                    }
```
Comment (sunng87) [inline src/catalog/src/information_schema/tables.rs:193]: would be nice if we can eliminate these checks

Comment (waynexia) [reply src/catalog/src/information_schema/tables.rs:193]: Same. We have to change `FrontendCatalogManager` to make it able to route a table that doesn't exist in the meta server.

File: src/catalog/src/local/memory.rs [hunk 7]
```diff
@@ -273,12 +281,13 @@ impl MemoryCatalogManager {
         }
     }
 
-    pub fn register_catalog_sync(&self, name: String) -> Result<bool> {
+    pub fn register_catalog_sync(self: Arc<Self>, name: String) -> Result<bool> {
```
Comment (zhongzc) [inline src/catalog/src/local/memory.rs:284]: Accepting `self: &Arc<Self>` as the argument can reduce tons of `self.catalogs **.clone()** .register_catalog_sync(...)`. Also for `create_catalog_entry`.

File: src/catalog/src/local/memory.rs [hunk 8]
```diff
@@ -273,12 +281,13 @@ impl MemoryCatalogManager {
...
         let mut catalogs = self.catalogs.write().unwrap();
 
-        match catalogs.entry(name) {
+        match catalogs.entry(name.clone()) {
             Entry::Vacant(e) => {
-                e.insert(HashMap::new());
+                let catalog = self.clone().create_catalog_entry(name.clone());
```
Comment (zhongzc) [inline src/catalog/src/local/memory.rs:289]:
```suggestion
                let catalog = self.create_catalog_entry(name.clone());
```

File: src/catalog/src/local/memory.rs [hunk 9]
```diff
@@ -332,9 +341,19 @@ impl MemoryCatalogManager {
         Ok(true)
     }
 
+    fn create_catalog_entry(self: Arc<Self>, catalog: String) -> SchemaEntries {
+        let information_schema = InformationSchemaProvider::build(
+            catalog.clone(),
```
Comment (zhongzc) [inline src/catalog/src/local/memory.rs:346]: Why clippy doesn't complain about unnecessary `clone`...

**Comment by zhongzc** [comment]
rest LGTM

File: src/catalog/src/local/memory.rs (hunk 9 above)
Comment (waynexia) [reply src/catalog/src/local/memory.rs:346]: good eyes... It took me two times to realize 🫣

>Why clippy doesn't complain about unnecessary clones...

I guess the other `catalog` makes things different. Let me check

File: src/catalog/src/local/memory.rs (hunk 7 above)
Comment (waynexia) [reply src/catalog/src/local/memory.rs]: makes sense, since this is not a trait method

**Review by zhongzc — COMMENTED** [review]
LGTM

**Comment by waynexia** [comment]
Count https://github.com/GreptimeTeam/greptimedb/pull/2149#pullrequestreview-1573120859 as a binding vote.

**Pull request merged**